### PR TITLE
Resolve region bounds from components of type projection

### LINKED ISSUE #121601: Borrow checker fails to pick up existing region bounds in the environment when aliases are given to assoicated types in trait bounds.
It occurs to me that adding named type terms to associated type of trait bounds can lead to borrow check failure, while the same code without "naming" the associated type passes the borrow check.

I tried this code:

```rust
pub trait Trait1 {
    type Output1;
    fn call<'z>(&'z self) -> &'z Self::Output1;
}

pub trait Trait2<T> {
    type Output2;
    fn call2<'x>(_: &'x T) -> &'x Self::Output2;
}

impl<A, B, T: Trait1<Output1 = A>> Trait2<T> for B // Mind this `A` here
where
    B: Trait2<A>,
{
    type Output2 = <B as Trait2<A>>::Output2;
    fn call2<'y>(source: &'y T) -> &'y Self::Output2 {
        let t = source.call();
        B::call2(t)
    }
}
```

I expected to see this happen:
Without naming the `Output1` to `A`, the borrow checker is happy.
```rust
pub trait Trait1 {
    type Output1;
    fn call<'z>(&'z self) -> &'z Self::Output1;
}

pub trait Trait2<T> {
    type Output2;
    fn call2<'x>(_: &'x T) -> &'x Self::Output2;
}

impl<B, T: Trait1> Trait2<T> for B
where
    B: Trait2<T::Output1>,
{
    type Output2 = <B as Trait2<T::Output1>>::Output2;
    fn call2<'y>(source: &'y T) -> &'y Self::Output2 {
        let t = source.call();
        B::call2(t)
    }
}
```

Instead, this happened:
```rust
error[E0309]: the parameter type `A` may not live long enough
  --> lifetime.rs:17:17
   |
16 |     fn call2<'y>(source: &'y T) -> &'y Self::Output2 {
   |              -- the parameter type `A` must be valid for the lifetime `'y` as defined here...
17 |         let t = source.call();
   |                 ^^^^^^^^^^^^^ ...so that the type `A` will meet its required lifetime bounds
   |
help: consider adding an explicit lifetime bound
   |
16 |     fn call2<'y>(source: &'y T) -> &'y Self::Output2 where A: 'y {
   |                                                      +++++++++++

error[E0309]: the parameter type `A` may not live long enough
  --> lifetime.rs:18:9
   |
16 |     fn call2<'y>(source: &'y T) -> &'y Self::Output2 {
   |              -- the parameter type `A` must be valid for the lifetime `'y` as defined here...
17 |         let t = source.call();
18 |         B::call2(t)
   |         ^^^^^^^^^^^ ...so that the type `A` will meet its required lifetime bounds
   |
help: consider adding an explicit lifetime bound
   |
16 |     fn call2<'y>(source: &'y T) -> &'y Self::Output2 where A: 'y {
   |                                                      +++++++++++

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0309`.
```

### Meta
This is found on the latest `master`, as of writing, and stable.

`rustc --version --verbose`:
```
rustc 1.78.0-nightly (256b6fb19 2024-02-06)
binary: rustc
commit-hash: 256b6fb19a2c018eaad4806d2369d1f6a71fc6ec
commit-date: 2024-02-06
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 17.0.6
```

<details><summary>Backtrace</summary>
<p>

Not an ICE, so not applicable.

</p>
</details>

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
Fix #121601

This will stay in draft so as to resolve potential issue in design and approach, possible conflict with existing work on the type checker.

### CODE REVIEW

**Comment by dingxiangfei2009** [comment]
:point_up: The existing build failure is a perfect example that I would like quote in a question about the well-formedness clauses which I am wondering about. I am formulating my question to be posted later in Zulip.

**Comment by compiler-errors** [comment]
This should be assigned to T-types when it exits draft mode

File: compiler/rustc_infer/src/infer/outlives/verify.rs [hunk 1]
```diff
@@ -56,6 +68,30 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
             }
         }
 
+        // A parameter or placeholder bound can also be aliased to an associated type
+        // so we need to inspect in the "reverse"
+        for &predicate in self.projection_predicates {
+            let Some(ty::ProjectionPredicate { term, projection_ty }) = predicate.no_bound_vars()
+            else {
+                continue;
```
Comment (dingxiangfei2009) [inline compiler/rustc_infer/src/infer/outlives/verify.rs:76]: ... so if we are here, we have, for now, `T == for<'a, 'b, ...> <... as ...>::...`.

I guess it is safe to say that `T: 'static`?

**Review by compiler-errors — CHANGES_REQUESTED** [review]
Thanks for the PR, but I'd personally lean towards closing this in favor of first having a discussion about whether we even want to support this at all.

Essentially this PR is a attempt to extend RFC 1214's projection rules (https://github.com/rust-lang/rfcs/blob/master/text/1214-projections-lifetimes-and-wf.md#outlives-for-projections) with a new rule:

```
TypeOutlivesViaBackwardsProjection:
  ∀i. R ⊢ Pi: 'a
  <P0 as Trait<P1..Pn>>::Assoc normalizes-to T
  --------------------------------------------------
  R ⊢ Ty: 'a
```

In order to fix code that looks like:

```rust
trait Foo: Bar {
    fn hello<'a>(&'a self) {}
}

trait Bar {
    type Assoc;
}

// impl<T> Foo for T where T: Bar // WORKS
impl<T, A> Foo for T where T: Bar<Assoc = A> // BROKEN
{
    fn hello<'a>(&'a self) {
        outlives::<'a, Self::Assoc>();
    }
}

fn outlives<'a, T: 'a>() {}
```

This is an attempt to fix a subset of a more general problem with the type system, where normalization only proceeds in one direction, even when considering an unnormalized type may allow us to prove goals that we cannot on the normalized type.

I don't believe this is unsound, as long as we're enforcing the same things that make RFC 1214's `OutlivesProjectionComponents` sound (i.e. projection must not give us unconstrained lifetimes).

However, I'm still generally concerned that this doesn't interact well with higher-ranked projection bounds, and is still quite limited in the exact matching it can do (consider, e.g., a chain of two projection bounds), and I don't know if it's worth the extra maintenance burden to fix this specific case, since I assume it's hit only very rarely and can be worked around.

---

Side-note: It reminds me of another limitation where we do not consider alias bounds when we are able to project to a rigid type, even if that rigid type doesn't have the same assumptions:

```rust
trait Foo {
    type Bar: Qux;
}

trait Qux {
    fn hello();
}

fn doesnt_work<T: Foo<Bar = U>, U>() {
    T::Bar::hello(); // <~ `T::Bar` becomes `U`, for which we know nothing.
}

fn works<T: Foo>() {
    T::Bar::hello();
}
```

**Comment by jackh726** [comment]
I think a good place to prototype this kind of change would be [a-mir-formality](https://github.com/rust-lang/a-mir-formality). I'm not sure if we're *quite* there yet, but that's where the types team wants to get to when thinking about changes that fundamentally interact with type and borrow checking.

File: compiler/rustc_infer/src/infer/outlives/verify.rs (hunk 1 above)
Comment (dingxiangfei2009) [reply compiler/rustc_infer/src/infer/outlives/verify.rs:76]: I retract. It is not necessarily.

**Pull request closed without merging**